scripts/build.py

```diff
@@ -715,62 +874,80 @@ def generate_build_directory():
     # /build folder. This is so that the replacing inside the HTML files works
     # correctly.
     hashes = dict()
-
-    # Create hashes for assets, copy directories and files to build/assets.
-    hashes.update(get_file_hashes(ASSETS_DEV_DIR))
-    copy_files_source_to_target(ASSETS_DEV_DIR, ASSETS_OUT_DIR, hashes)
-
-    # Process third_party resources, create hashes for them and copy them into
-    # build/third_party/generated.
+    build_tasks = collections.deque()
+    copy_tasks = collections.deque()
+    # Process third_party resources, and copy them into staging directory.
     build_minified_third_party_libs()
-    hashes.update(get_file_hashes(THIRD_PARTY_GENERATED_STAGING_DIR))
-    copy_files_source_to_target(
-        THIRD_PARTY_GENERATED_STAGING_DIR,
-        THIRD_PARTY_GENERATED_OUT_DIR, hashes)
-
-    # Minify extension static resources, create hashes for them and copy them
-    # into build/extensions.
-    hashes.update(get_file_hashes(EXTENSIONS_DEV_DIR))
-    build_files(
-        EXTENSIONS_DEV_DIR, EXTENSIONS_STAGING_DIR, hashes)
-    copy_files_source_to_target(
-        EXTENSIONS_STAGING_DIR, EXTENSIONS_OUT_DIR, hashes)
-
-    # Create hashes for all template files.
-    hashes.update(get_file_hashes(TEMPLATES_DEV_DIR_CORE))
 
+    # Create hashes for all directories and files.
+    HASH_DIRS = [
+        ASSETS_DEV_DIR, EXTENSIONS_DIR.get('dev_dir'),
+        TEMPLATES_CORE_DIR.get('dev_dir'), THIRD_PARTY_GENERATED_STAGING_DIR]
+    for HASH_DIR in HASH_DIRS:
+        hashes.update(get_file_hashes(HASH_DIR))
     # Save hashes as JSON and write the JSON into JS file
     # to make the hashes available to the frontend.
-    save_hashes_as_json(HASHES_JSON, hashes)
-
+    if not does_directory_tree_exist(HASHES_JSON):
+        os.makedirs(os.path.dirname(HASHES_JSON))
+    with open(HASHES_JSON, 'w+') as hashes_js_file:
+        write_to_file_stream(
+            hashes_js_file, get_hashes_json_file_contents(hashes))
+    # Update hash dict with newly created hashes.js.
+    hashes.update({HASHES_JS_FILENAME: generate_md5_hash(HASHES_JSON)})
```

Comment:
Do we actually need to create a hash of hashes.js? /cc @vojtechjelinek

Reply:
`hashes.js` has to reside within `/build`. I think hashing `hashes.js` is also a part of the cache invalidation effort.

Reply:
In the old version of the code, where did we do this? I can't find it, but I might just be missing it.